File: utils/parametric_virtual_miking.py
import numpy as np
from icecream import ic
from utils.localize_sources import localize_sources
from utils.define_spherical_p import define_spherical_p
from utils.define_dereverb_params import define_dereverb_params
from lib.dereverb_array_with_doa import dereverb_array_with_doa
from lib.spherical_harmonics_estimation import spherical_harmonics_estimation
from lib.estimate_direct_signal import estimate_direct_signal

"""
"
"""


def parametric_virtual_miking(macro, params, sources, array, cpt_pts):
    print("PARAMETRIC VIRTUAL MIKING")

    # Localize the sources using ransac
    sources = localize_sources(macro, params, sources, array)

    # Remove reverberation from the microphone signals
    dereverb_params = define_dereverb_params(array)
    array = dereverb_array_with_doa(macro, params, sources, array, dereverb_params)

    # Spherical Parameters
    spherical_p = define_spherical_p(sources, array)

    # Harmonics coefficients
    h_coeff = spherical_harmonics_estimation(macro, params, sources, array, spherical_p)

    # Estimate the direct signal using the sph expansion
    cpt_pts = estimate_direct_signal(macro, params, sources, array, cpt_pts, h_coeff, spherical_p)

    # @ line 103

    return cpt_pts

File: parametric_vm/parametric_virtual_miking.m
function cptPts = parametric_virtual_miking(array, source, cptPts, params, macro, quickload3, ...
                                          quickload4, quickload5, quickload6)
    %% parametricvirtualmiking
    % This function performs virtual miking using a parametric technique.
    % Parameters:
    %   array: struct containing information on the distributed arrays.
    %   source: struct containing information on the sources.
    %   cptPts: struct containing information on the VMs.
    %
    % Mirco Pezzoli, 2021
    % 22/06/2021
    % v. 0.1

    %% Localize the sources
    source = find_source_position(array, source, params, macro, quickload3);

    % Remove reverberation from the microphone signals
    array = remove_reverb(array, source, params, macro, quickload4);

    %% Spherical harmonics expansion estimation
    sphParams = define_spherical_params(source, array);
    hCoeff = get_hCoeff(array, source, sphParams, params, macro, quickload5);

    %% Estimate the direct signal using the sph expansion
    cptPts = get_direct_signal(cptPts, hCoeff, array, source, sphParams, params, macro, quickload6);

    %% Estimate full signal with the diffuse component
    cptPts = estimate_complete_signal(cptPts, array, params);

end


%    cptPts = S.cptPts;
%    array = S.array;